salt/modules/zypperpkg.py

```diff
@@ -2064,56 +2132,67 @@ def remove_lock(packages, root=None, **kwargs):  # pylint: disable=unused-argume
 
 def hold(name=None, pkgs=None, **kwargs):
     '''
-    Add a package lock. Specify packages to lock by exact name.
+    Add a package hold.  Specify one of ``name`` and ``pkgs``.
 
-    root
-        operate on a different root directory.
+    name
+        A package name, or a comma-separated list of package names.  Specify
+        one of ``name`` or ``pkgs``.
+
+    pkgs
+        A list of packages.  Specify one of ``name`` or ``pkgs``.
 
     CLI Example:
 
     .. code-block:: bash
 
-        salt '*' pkg.add_lock <package name>
-        salt '*' pkg.add_lock <package1>,<package2>,<package3>
-        salt '*' pkg.add_lock pkgs='["foo", "bar"]'
-
-    :param name:
-    :param pkgs:
-    :param kwargs:
-    :return:
+        salt '*' pkg.hold <package name>
+        salt '*' pkg.hold <package1>,<package2>,<package3>
+        salt '*' pkg.hold pkgs='["foo", "bar"]'
     '''
     ret = {}
     root = kwargs.get('root')
-    if (not name and not pkgs) or (name and pkgs):
-        raise CommandExecutionError('Name or packages must be specified.')
-    elif name:
-        pkgs = [name]
+    if (not name and not pkgs):
+        raise SaltInvocationError('Name or packages must be specified.')
 
     locks = list_locks(root=root)
     added = []
     try:
-        pkgs = list(__salt__['pkg_resource.parse_targets'](pkgs)[0].keys())
+        pkgs = list(__salt__['pkg_resource.parse_targets'](name, pkgs)[0].keys())
     except MinionError as exc:
         raise CommandExecutionError(exc)
 
     for pkg in pkgs:
         ret[pkg] = {'name': pkg, 'changes': {}, 'result': False, 'comment': ''}
         if not locks.get(pkg):
             added.append(pkg)
-            ret[pkg]['comment'] = 'Package {0} is now being held.'.format(pkg)
+            if __opts__['test']:
+                ret[pkg].update(result=None, comment='Package {0} is set to be held.'.format(pkg))
+            else:
+                ret[pkg].update(result=True, changes={'new': 'hold', 'old': ''},
+                                comment='Package {0} is now being held.'.format(pkg))
         else:
             ret[pkg]['comment'] = 'Package {0} is already set to be held.'.format(pkg)
+            ret[pkg].update(result=True)
 
-    if added:
+    if added and not __opts__['test']:
         __zypper__(root=root).call('al', *added)
+        if __zypper__.exit_code != 0:
```

Comment:
enough just `if __zypper__.exit_code:` but then you need to rather use `defaults.exitcodes.EX_OK`.